File: packs/musicvideo/nexgen_pack_musicvideo/patterns_similarity.py
"""Pattern-Aehnlichkeits-Matrix (v0.13.0).

User-Anfrage "aehnlich zu Shinkai aber mit mehr Aktion": der Skill
liefert die naechsten 3-5 Nachbarn eines Pattern.

Aehnlichkeit = gewichteter Mittelwert aus:
- Cosine-Similarity der framing_mix-Vektoren (Gewicht 0.5)
- Jaccard-Similarity der camera_vocabulary-Sets (Gewicht 0.3)
- Abstand der asl_range.typical_s (Gewicht 0.2, normalisiert auf
  log-Skala, weil ASL exponentiell variiert)

Keine ML, kein Training — pure Vektor-Mathe ueber strukturierte
Pattern-Felder.
"""

from __future__ import annotations

import math
import re

from nexgen_pack_musicvideo.patterns_schema import Pattern, load_all_patterns
from nexgen_engine.shotlist.schema import Framing


_FRAMINGS_ORDER = (
    Framing.WIDE, Framing.FULL, Framing.MS, Framing.MCU, Framing.CU,
    Framing.ECU, Framing.OTS, Framing.POV, Framing.INSERT, Framing.AERIAL,
)


def _framing_vector(p: Pattern) -> list[float]:
    mix = p.framing_mix.by_framing()
    return [float(mix[f]) for f in _FRAMINGS_ORDER]


def _cosine(a: list[float], b: list[float]) -> float:
    if len(a) != len(b):
        return 0.0
    dot = sum(x * y for x, y in zip(a, b))
    na = math.sqrt(sum(x * x for x in a))
    nb = math.sqrt(sum(y * y for y in b))
    if na == 0 or nb == 0:
        return 0.0
    return dot / (na * nb)


_WORD_RE = re.compile(r"[a-zA-Z]+")


def _camera_token_set(p: Pattern) -> set[str]:
    """Token-Set aus camera_vocabulary, fuer Jaccard."""
    tokens: set[str] = set()
    for entry in p.camera_vocabulary:
        for w in _WORD_RE.findall(entry.lower()):
            if len(w) >= 3:
                tokens.add(w)
    return tokens


def _jaccard(a: set[str], b: set[str]) -> float:
    if not a and not b:
        return 0.0
    return len(a & b) / max(len(a | b), 1)


def _asl_log_distance(a: float, b: float) -> float:
    """Distanz auf log-Skala, normalisiert auf [0, 1]. Naeher = besser."""
    if a <= 0 or b <= 0:
        return 0.0
    diff = abs(math.log10(a) - math.log10(b))
    # log10-Differenz von 1 = Faktor 10 = sehr entfernt.
    return max(0.0, 1.0 - min(1.0, diff))


def similarity(a: Pattern, b: Pattern) -> float:
    """Gewichteter Aehnlichkeits-Score zwischen zwei Pattern in [0, 1]."""
    cos = _cosine(_framing_vector(a), _framing_vector(b))
    jac = _jaccard(_camera_token_set(a), _camera_token_set(b))
    asl = _asl_log_distance(a.asl_range.typical_s, b.asl_range.typical_s)
    return 0.5 * cos + 0.3 * jac + 0.2 * asl


def suggest_similar(
    pattern_id: str, *, top: int = 5
) -> list[tuple[Pattern, float]]:
    """Liefert die Top-N aehnlichsten Pattern zu einem Anker-Pattern.

    User-Anfrage "aehnlich zu Shinkai" → Skill nutzt diese Funktion
    und zeigt 3-5 Nachbarn mit Score-Anzeige.
    """
    library = load_all_patterns()
    anchor = next((p for p in library if p.id == pattern_id), None)
    if anchor is None:
        return []
    scored: list[tuple[Pattern, float]] = []
    for p in library:
        if p.id == anchor.id:
            continue
        s = similarity(anchor, p)
        scored.append((p, s))
    scored.sort(key=lambda ps: ps[1], reverse=True)
    return scored[:top]

File: Sources/NexGenEngine/Packs/Musicvideo/PatternsSimilarity.swift
import Foundation

/// Pattern-similarity matrix (v0.13.0). Port of
/// `nexgen_pack_musicvideo/patterns_similarity.py`.
///
/// A user request like "similar to Shinkai but with more action" is served
/// by returning the nearest 3-5 neighbors of a pattern.
///
/// Similarity = weighted average of:
/// - cosine similarity of the `framingMix` vectors (weight 0.5)
/// - Jaccard similarity of the `cameraVocabulary` token sets (weight 0.3)
/// - distance of `aslRange.typicalS` (weight 0.2, normalized on a log scale,
///   since ASL varies exponentially)
///
/// No ML, no training — pure vector math over structured pattern fields.
public enum PatternsSimilarity {
    private static let framingsOrder: [Framing] = [
        .wide, .full, .ms, .mcu, .cu, .ecu, .ots, .pov, .insert, .aerial,
    ]

    private static func framingVector(_ p: Pattern) -> [Double] {
        let mix = p.framingMix.byFraming()
        return framingsOrder.map { Double(mix[$0] ?? 0) }
    }

    private static func cosine(_ a: [Double], _ b: [Double]) -> Double {
        guard a.count == b.count else { return 0.0 }
        let dot = zip(a, b).reduce(0.0) { $0 + $1.0 * $1.1 }
        let na = (a.reduce(0.0) { $0 + $1 * $1 }).squareRoot()
        let nb = (b.reduce(0.0) { $0 + $1 * $1 }).squareRoot()
        guard na != 0, nb != 0 else { return 0.0 }
        return dot / (na * nb)
    }

    private static let wordPattern = try! NSRegularExpression(pattern: "[a-zA-Z]+")

    /// Token set from `cameraVocabulary`, for Jaccard.
    private static func cameraTokenSet(_ p: Pattern) -> Set<String> {
        var tokens: Set<String> = []
        for entry in p.cameraVocabulary {
            let lower = entry.lowercased()
            let ns = lower as NSString
            for match in wordPattern.matches(in: lower, range: NSRange(location: 0, length: ns.length)) {
                let word = ns.substring(with: match.range)
                if word.count >= 3 { tokens.insert(word) }
            }
        }
        return tokens
    }

    private static func jaccard(_ a: Set<String>, _ b: Set<String>) -> Double {
        guard !a.isEmpty || !b.isEmpty else { return 0.0 }
        return Double(a.intersection(b).count) / Double(max(a.union(b).count, 1))
    }

    /// Distance on a log scale, normalized to [0, 1]. Closer = better.
    private static func aslLogDistance(_ a: Double, _ b: Double) -> Double {
        guard a > 0, b > 0 else { return 0.0 }
        let diff = abs(log10(a) - log10(b))
        // A log10 difference of 1 = factor 10 = very far apart.
        return max(0.0, 1.0 - min(1.0, diff))
    }

    /// Weighted similarity score between two patterns in [0, 1]. Port of
    /// `patterns_similarity.py::similarity`.
    public static func similarity(_ a: Pattern, _ b: Pattern) -> Double {
        let cos = cosine(framingVector(a), framingVector(b))
        let jac = jaccard(cameraTokenSet(a), cameraTokenSet(b))
        let asl = aslLogDistance(a.aslRange.typicalS, b.aslRange.typicalS)
        return 0.5 * cos + 0.3 * jac + 0.2 * asl
    }

    /// Returns the top-N most similar patterns to an anchor pattern.
    ///
    /// A user request like "similar to Shinkai" uses this function and shows
    /// 3-5 neighbors with score display. Port of
    /// `patterns_similarity.py::suggest_similar`.
    public static func suggestSimilar(patternId: String, top: Int = 5) throws -> [(pattern: Pattern, score: Double)] {
        let library = try Patterns.loadAllPatterns()
        guard let anchor = library.first(where: { $0.id == patternId }) else { return [] }
        var scored: [(Pattern, Double)] = []
        for p in library where p.id != anchor.id {
            scored.append((p, similarity(anchor, p)))
        }
        scored.sort { $0.1 > $1.1 }
        return Array(scored.prefix(top))
    }
}
